src/backup.cpp

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2020 YADRO
+
+#include "accounts.hpp"
+#include "backup.hpp"
+#include "manifest.hpp"
+
+#include <vector>
+
+namespace fs = std::filesystem;
+
+// clang-format off
+/** @brief List of base configuration files/directories. */
+static const std::vector<const char*> baseConfigs = {
+    "etc/dropbear/dropbear_rsa_host_key",
+    "etc/ipmi_pass",
+    "etc/machine-id",
+};
+
+/** @brief List of network configuration files/directories. */
+static const std::vector<const char*> networkConfigs = {
+    "etc/hostname",
+    "etc/systemd/network",
+    "var/lib/first-boot-set-hostname",
+};
+// clang-format on
+
+Backup::Backup(const Configuration& cfg) : config(cfg)
+{
+    std::string tmpPathTemplate = fs::temp_directory_path() / "backup_XXXXXX";
+    tmpDir = mkdtemp(tmpPathTemplate.data());
+}
+
+Backup::~Backup()
+{
+    fs::remove_all(tmpDir);
+}
+
+void Backup::backup() const
+{
+    if (config.handleAccounts)
+    {
+        Accounts acc(config.rootFs, tmpDir, config.readOnlyFs);
+        acc.backup();
+    }
+
+    for (const auto& it : baseConfigs)
+    {
+        backupFile(it);
+    }
+
+    if (config.handleNetwork)
+    {
+        for (const auto& it : networkConfigs)
+        {
+            backupFile(it);
+        }
+    }
+
+    const Manifest manifest = Manifest::fromOsRelease(config.rootFs);
```

Comment:
I'd say that manifest should also contain the information about the hostname, and that shall be displayed to the user during restoration with a confirmation question.